File: cpp/_141_Linked_List_Cycle.cpp
/*
Given a linked list, determine if it has a cycle in it.

Follow up:
Can you solve it without using extra space?
*/
#include <iostream>

using namespace std;

class Solution {
    

    struct ListNode {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(NULL) {}
    };

    public:
        bool hasCycle(ListNode *head) {
            if(!head) return false;
            ListNode *walker = head;
            ListNode *runner = head;
            while(runner->next != NULL && runner->next->next != NULL) {
                walker = walker->next;
                runner = runner->next->next;
                if(walker->val == runner->val) {
                    return true;
                }
            }
            return false;
        }
};

int main() {
    Solution sol;
    return 0;
}

/*
https://en.wikipedia.org/wiki/Cycle_detection

*/

File: py/_141_Linked_List_Cycle.py
"""
Given a linked list, determine if it has a cycle in it.

Follow up:
Can you solve it without using extra space?
"""

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slow = head
        fast = head
        while fast and fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
            if slow.val == fast.val:
                return True
        return False
